### ISSUE TITLE
never_loop false positive

### ISSUE THREAD

**Issue by crumblingstatue** [Manishearth/rust-clippy#1586]
```rust
fn main() {
	let mut i = 0;
	loop {
		if i < 10 {
			i += 1;
			println!("Iteration {}", i);
			continue;
		}
		return;
	}
}
```

```
warning: this loop never actually loops
  --> src/main.rs:3:2
   |
3  |   	loop {
   |  __^ starting here...
4  | | 		if i < 10 {
5  | | 			i += 1;
6  | | 			println!("Iteration {}", i);
7  | | 			continue;
8  | | 		}
9  | | 		return;
10 | | 	}
   | |__^ ...ending here
   |
   = note: #[warn(never_loop)] on by default
   = help: for further information visit https://github.com/Manishearth/rust-clippy/wiki#never_loop


```

Also the wiki entry for never_loop is missing.

**Comment by mcarton** [comment]
> Also the wiki entry for never_loop is missing.

We should deprecate the wiki and use https://manishearth.github.io/rust-clippy/current/ at some point. Cc @Manishearth, @oli-obk, @llogiq, @killercup

**Comment by crumblingstatue** [comment]
At a simple glance, it seems to me that never_loop is actually very naively implemented. It basically just looks for statements that unconditionally terminate the loop, but disregards `continue` completely.

It might need to do a more complicated control flow analysis if it wants to be an accurate lint.

**Comment by crumblingstatue** [comment]
Here's another "interesting" example, using labelled loops:

```rust
fn main() {
	'outer: loop {
		println!("I always loop");
		loop {
			println!("I indeed never loop");
			continue 'outer; // false negative, this unconditionally redirects control flow
			                 // and ends this inner loop
		}
		break; // false positive, unreachable
	}
}

```

**Comment by camsteffen** [comment]
Here's some more examples with `match`:
```rust
loop {
    match 'a' {
        'a' => continue,
        _ => (),
    }
    break
}

loop {
    return match 'a' {
        'a' => continue,
        _ => (),
    }
}
```

**Comment by birkenfeld** [comment]
Another instance in `walkdir`, with the structure:
```
loop {
    if pred() {
        continue;
    }
    return;
}
```
The lint should probably be disabled until it is either removed or made more aware of continues.

**Comment by camsteffen** [comment]
I'd like to try and fix this

**Comment by mcarton** [comment]
@camsteffen 👍  Go ahead! Don't hesitate to ask if you have any question or to open a PR with a partial fix to get some review before you have finished.

**Comment by uHOOCCOOHu** [comment]
It seem that #1804 doesn't fix it completely. This example mentioned above still causes the warning. 
```rust
loop {
    return match 'a' {
        'a' => continue,
        _ => (),
    }
}
```
@camsteffen

**Comment by oli-obk** [comment]
Thanks for the report and fix @uHOOCCOOHu
